File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
def solution(gems):
    gem = list(set(gems))
    n = len(gem)
    
    dic = {gems[0]: 1}
    answer = [1, len(gems)]
    left,right = 0,0
    while left <= right and right < len(gems):
        if len(dic) == n:
            if answer[1]-answer[0] > right-left:
                answer = [left+1, right+1]
            
            dic[gems[left]] -= 1
            if dic[gems[left]] == 0:
                del dic[gems[left]]
            left += 1

        elif len(dic) < n:
            right += 1
            if right >= len(gems):
                break
            if gems[right] not in dic:
                dic[gems[right]] = 1
            else:
                dic[gems[right]] += 1
        
    return answer

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.js
function solution(gems) {
    var answer = [0, gems.length-1];
    let gemKinds = new Set(gems).size
    let gemMap = new Map()
    
    let start = 0
    for(let end = 0; end < gems.length; end++){
        
        // 보석 넣기
        gemMap.set(gems[end], (gemMap.get(gems[end])||0) + 1)
        
        // 종류가 다 있는 경우
        while(gemMap.size == gemKinds){
            
            // 최소 갱신
            if(end - start < answer[1] - answer[0]){
                answer = [start, end]
            }
            
            gemMap.set(gems[start],(gemMap.get(gems[start]) - 1 ))
            if(gemMap.get(gems[start])===0) gemMap.delete(gems[start])
            start++;
        }
    }
    return [answer[0]+1, answer[1]+1]
}